openmls/src/credentials/mod.rs

```diff
@@ -81,9 +83,81 @@ impl TryFrom<u16> for CredentialType {
 ///
 /// This struct contains an X.509 certificate chain.  Note that X.509
 /// certificates are not yet supported by OpenMLS.
-#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+#[derive(
+    Debug, PartialEq, Clone, Serialize, Deserialize, TlsDeserialize, TlsSerialize, TlsSize,
+)]
 pub struct Certificate {
-    cert_data: Vec<u8>,
+    identity: TlsByteVecU16,
+    cert_chain: TlsVecU16<TlsByteVecU16>,
+}
+
+impl Certificate {
+    fn parse(&self) -> Result<Vec<X509Certificate>, CredentialError> {
+        self.cert_chain.iter().try_fold(
+            Vec::new(),
+            |mut acc, certificate| -> Result<Vec<X509Certificate>, CredentialError> {
+                acc.push(Self::parse_single(certificate)?);
+                Ok(acc)
+            },
+        )
+    }
+
+    fn parse_single(certificate: &TlsByteVecU16) -> Result<X509Certificate, CredentialError> {
+        x509_parser::parse_x509_certificate(certificate.as_slice())
+            .map(|(_, cert)| cert)
+            .map_err(CredentialError::from)
+    }
+
+    /// Is signed by issuer
+    fn is_verified(
+        certificate: &X509Certificate,
+        issuer: &X509Certificate,
+    ) -> Result<(), CredentialError> {
+        certificate
+            .verify_signature(Some(&issuer.subject_pki))
+            .map_err(|_| CredentialError::InvalidCertificateChain)
+    }
+
+    fn is_valid<'a>(
+        certificate: &'a X509Certificate,
+    ) -> Result<&'a X509Certificate<'a>, CredentialError> {
+        if Self::is_time_valid(certificate) && Self::is_structure_valid(certificate) {
+            Ok(certificate)
+        } else {
+            Err(CredentialError::InvalidCertificate)
+        }
+    }
+
+    fn is_structure_valid(certificate: &X509Certificate) -> bool {
+        // validates structure (fields etc..)
+        struct NoopLogger;
+        impl Logger for NoopLogger {
+            fn warn(&mut self, _: &str) {}
+            fn err(&mut self, _: &str) {}
+        }
+        X509StructureValidator.validate(certificate, &mut NoopLogger)
+    }
+
+    fn is_time_valid(certificate: &X509Certificate) -> bool {
+        // 'not_before' < now < 'not_after'
+        certificate.validity().is_valid()
+    }
+
+    fn signature_scheme(_certificate: &X509Certificate) -> SignatureScheme {
+        // TODO: (wire) implement x509 properly. Should be derived from issuer public key info
+        SignatureScheme::ED25519
```

Comment:
Might be able to get it from here https://docs.rs/x509-parser/latest/x509_parser/certificate/struct.TbsCertificate.html#structfield.signature